### DISCUSSION TITLE
Rayleigh Benard convection

### DISCUSSION THREAD

**Discussion by iszoteptec** [dynamicslab/pysindy#460] [General]
Hello all,

In general, I wondered if anyone has had any success in applying SINDy to 2D or 3D Rayleigh-Benard convection?

More specifically, I have observed that in fitting the individual terms of this problem, the weak formulation can obtain not only 'swapped' terms that are relatable through the incompressibility condition (typically for the diffusivity terms), but also swapped advection terms that don't appear to relate, but do produce a similar or better mean-squared error in comparison between the training integral and the model fit.

Has anyone else observed this kind of behaviour? Is it common with the weak formulation? I am relatively new to ML methods (<1yr application time).

Using physical knowledge to constrain the model can make the fit far more robust, but it would be ideal to be able to reliably obtain a robust fitting recipe without such constraints, so as to apply the same recipe to an extended problem where the contraints are less clear.

For reference, I have:-
1. Adjusted the hyperparameter to optimise the process
2. Added more data to explore the extent to which we may be in a low data limit
3. Ensembled
4. Checked errors with comparison to training integrals
5. Used physical knowledge to obtain robust fits, but prefer to also obtain robust unconstrained fitting

Thanks in advance for any advice.
Chris W (University of Leeds, UK)

**Comment by Jacob-Stevens-Haas** [comment]
Hi Chris, it sounds like a cool problem!  The answer is probably yes - it appears in [this repo](https://github.com/williamgilpin/dysts/blob/b42f88b32c35d10a6839d2c6e67e32002ed26e89/dysts/flows.py#L909), and the author of that repo/associated papers has used SINDy.  I don't know if anyone's looked at it in detail, or whether that equation is the trajectory of an individual particle or a  low-dimensional parametrization of a PDE.

**Reply by iszoteptec** [reply]
Thanks Jacob, much appreciated. I'll explore that repo and the papers.

**Comment by iszoteptec** [comment]
All,
For completeness and anyone searching for a similar problem, I tracked this down fairly quickly to a compatability issue between the 1.7.5 pip install of SINDy and a 1.7.5 version downloaded from the master branch at a later date (slightly unknown, probably September 2023). Anyway, the solution was to use the later downloaded version and this advection term 'swap' problem went away - probably related to how we were defining our library, ordering and then constraining coefficients within the library. The terms were in effect identical, but perhaps ordering and model.print was showing up this confusion at print command time.
Chris

**Reply by Jacob-Stevens-Haas** [reply]
Yes, the current master is tracking for a 2.0 release (aka backwards-incompatible) once we clean up the trapping PRs in the next month or so.  We hope to improve the ease and clarity of constraint specification in a future release.

**Reply by iszoteptec** [reply]
Ahh, good to know, thanks!
